### ISSUE TITLE
Luvit won't report loop error when loading module 

### ISSUE THREAD

**Issue by Miigon** [luvit/luvit#1049]
## Description
Basically, when two Lua files are trying to require each other, an error **should** occur. (that's the original behavior of Lua)
But that doesn't happen in luvit. Luvit just returns `{ }` when the loop require occurs.
### Code
foo.lua:
```lua
local bar = require "./bar"

return {"this is foo!",bar = bar}
```

bar.lua:
```lua
local foo = require "./foo"

return {"this is bar",foo = foo}
```

test.lua:
```lua
local foo = require "./foo"
local bar = require "./bar"

print "foo:"
p(foo)
print "bar:"
p(bar)
```
### Expected Behavior:
```
$ luajit test.lua
luajit: .///bar.lua:1: loop or previous error loading module './foo'
stack traceback:
	[C]: in function 'require'
	.///bar.lua:1: in main chunk
	[C]: in function 'require'
	.///foo.lua:1: in main chunk
	[C]: in function 'require'
	test.lua:1: in main chunk
	[C]: at 0x0100001120
```
### Actual Behavior:
```
$ luvit test
foo:
{ 'this is foo!', bar = { 'this is bar', foo = { } } }
bar:
{ 'this is bar', foo = { } }
```
notice `foo = { }`. This shouldn't be the case. Instead, luvit should simply throw an error (as original lua does).
## Fix
This is something to do with `module:require()` and `moduleCache` in `deps/require.lua`. And could be fixed by adding a local table `moduleLoading` to `deps/require.lua` to check whether a module is currently loading or not.

I think this is not a very serious problem, it can sometimes be confusing though. If you guys agree, I could open a PR to fix it and maybe also add a test. :)

**Comment by Miigon** [comment]
Btw, I actually ran into loop require problems (two modules requiring each other) in my own project. I want to know if anyone has a good idea to solve the problem.

**Comment by SinisterRectus** [comment]
If you're running into a require cycle, then you may want to reconsider your module design.

If you absolutely cannot do this, you can foward declare the module and require it when it's needed.

```lua
local foo = {}

local bar

function foo.load()
  bar = require("./bar")
end

return foo
```

Luvit probably should throw an error for this, but I'm not currently familiar enough with the require code to go and change it. This may also strictly be a breaking change.

Have you tried this with luvit-loader?

**Comment by SinisterRectus** [comment]
Interesting. I convinced myself that Luvit's circular requires do not work because they weren't supposed to work.

**Comment by squeek502** [comment]
Yeah, I agree that allowing circular requires is weird--I was surprised to see that test when I was porting the `require` tests to `luver`.

From what I can tell, circular requires are not really possible without the `exports` table, as I believe this is what's happening:

- `parent` gets required, and has its environment set so that an `exports` table exists (set to `module.exports` in luvit's require cache)
- `parent` modifies the exports table, which is a reference to `module.exports` in require, so it gets updated immediately
- `parent` requires `child`, which in turn requires `parent`, and because of the above, it receives a direct reference to the `parent` exports table, so any changes made to `exports` also get handed to `child`

With the `return` structure, though, `parent` has to finish executing (i.e. the function returned from `loadstring` has to actually be fully run) before the result is acquired, so there's no chance for `child` to get access to it before it exists. I guess there might be some hackery that could be done here to make it work anyway (like stuffing the results into the `exports` table? I haven't thought it all the way through), but I'm not sure.

Anyway, circular requires have been broken ever since Luvit moved away from the `exports` table ([Jan 2016](https://github.com/luvit/luvit/commit/41bb62a13475ba0ad39d851df2cd4ca925cc30ff)), so maybe it's somewhat safe to assume no one is still using them? Would it be okay to stop supporting circular requires? @creationix, thoughts?

**Comment by aiverson** [comment]
I would rather not allow circular requires. I think it is an architectural flaw to use a circular require or dependency.

**Comment by Miigon** [comment]
@squeek502 And it is working only when you put the require statement of `child` AFTER all the function definitions. Otherwise `child` wouldn't see those functions. But probably no one is writing codes like that.

**Comment by creationix** [comment]
Circular requires were implemented because originally luvit was a port of node.js where it's explicitly supported.

**Comment by aiverson** [comment]
In my opinion, circular requires are a footgun. They shouldn't be necessary anywhere and they are a place where it is particularly easy to mess up. Having a nice error message and failing loudly is the best way to handle it.
